src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java

```diff
@@ -151,10 +159,12 @@ static SqlAuthenticationToken getSqlFedAuthTokenPrincipal(SqlFedAuthInfo fedAuth
                                                                     : fedAuthInfo.spn + defaultScopeSuffix;
         Set<String> scopes = new HashSet<>();
         scopes.add(scope);
-
-        lock.lock();
-
+        
+        boolean lockAcquired = false;
         try {
+        	//Just try to acquire the lock and if can't then proceed to attempt to get the token
+            lockAcquired = lock.tryLock(Math.min(millisecondsRemaining, TOKEN_LOCK_WAIT_DURATION_MS), TimeUnit.MILLISECONDS);
```

Comment:
Where are we using `lockAcquired`? We should go ahead only if it returns true(lock acquired), right?

Reply:
The idea is to attempt acquiring the lock with a timeout of 5 seconds. If the lock cannot be acquired within this time, proceed to attempt token acquisition regardless of whether the earlier thread(s) have released the lock or not

`lockAcquired` is used to check in the finally block if the lock should be released or not.

Reply:
The purpose is to optimize the token acquisition process, the first caller succeeding does caching which is then leveraged by subsequent threads.  However, if the first thread takes considerable time, then we want the others to also go and try after waiting for a while.

If we were to let say 30 threads try in parallel, they would all miss the cache and hit the AAD auth endpoints to get their tokens at the same time, stressing the auth endpoint.

Reply:
Understood the intent, but would Semaphore be a more readable solution?

Reply:
Agree. The code has been updated along with detailed comment describing the intent.